bionic/core/task_execution.py

```diff
@@ -695,50 +768,71 @@ def _get_digests_by_dnode(self):
+    of their follow-up tasks (which need to be computed immediately after their
+    parents). It also maintains a copy of this subgraph with unnecessary data pruned;
+    these "stripped" TaskStates can be safely serialized with cloudpickle and sent to
+    the other process.
+    """
 
-        Parameters
-        ----------
+    def __init__(self, target_state, bootstrap):
+        self._target_state = target_state
+        self._bootstrap = bootstrap
 
-        new_task_states_by_key: Dict from key to stripped states, optional, default is ``{}``
-            The cache for tracking stripped task states.
-        """
+        self._stripped_states_by_task_key = {}
+        self.persistable_but_not_persisted_states = set()
 
-        if new_task_states_by_key is None:
-            new_task_states_by_key = {}
+        self._strip_state(target_state)
 
-        # All task keys should point to the same task state.
-        if self.task_keys[0] in new_task_states_by_key:
-            return new_task_states_by_key[self.task_keys[0]]
+    def _strip_state(self, original_state):
+        task_key = original_state.task.keys[0]
+        if task_key in self._stripped_states_by_task_key:
+            return self._stripped_states_by_task_key[task_key]
+        original_state.set_up_caching_flags(self._bootstrap)
+        stripped_state = copy.copy(original_state)
+        self._stripped_states_by_task_key[task_key] = stripped_state
 
-        # Let's make a copy of the task state.
-        # This is not a deep copy so we'll avoid mutating any of the member variables.
-        task_state = copy.copy(self)
+        # Strip out data cached in memory -- we can't necessarily pickle it, so
+        # we need to get rid of it before trying to transmit this state to another
+        # process.
+        stripped_state._results_by_dnode = None
 
-        # Clear up memoized cache to avoid sending it through IPC.
-        task_state._results_by_dnode = None
-        # Clear up fields not needed in subprocess, for computing or for cache lookup.
-        task_state.func_attrs = None
-        task_state.entity_defs_by_dnode = None
-        task_state.case_key = None
-        task_state._provenance = None
+        # These fields are picklable, but only needed for cache setup and
+        # initialization.
+        if stripped_state.is_initialized:
+            stripped_state.func_attrs = None
+            stripped_state.entity_defs_by_dnode = None
+            stripped_state.case_key = None
 
-        if task_state.is_cached:
-            task_state.dep_states = []
-            task_state.task = None
-        else:
-            task_state.dep_states = [
-                dep_state.strip_state_for_subprocess(new_task_states_by_key)
-                for dep_state in task_state.dep_states
-            ]
+        if stripped_state.should_persist:
+            assert len(stripped_state.followup_states) == 0
```

Comment:
Yes, it does seem pretty important. I feel like it's valuable to have the assertion here because it's related to work we're doing, but I'd be happy to also put it in `set_up_caching_flags`.